packages/openapi-v3/src/controller-spec.ts

```diff
@@ -214,47 +209,84 @@ function resolveControllerSpec(constructor: Function): ControllerSpec {
 
     for (const p of paramTypes) {
       if (isComplexType(p)) {
-        if (!spec.components) {
-          spec.components = {};
-        }
-        if (!spec.components.schemas) {
-          spec.components.schemas = {};
-        }
-        if (p.name in spec.components.schemas) {
-          // Preserve user-provided definitions
-          debug(
-            '    skipping parameter type %j as already defined',
-            p.name || p,
-          );
-          continue;
-        }
-        const jsonSchema = getJsonSchema(p);
-        const openapiSchema = jsonToSchemaObject(jsonSchema);
-        const outputSchemas = spec.components.schemas;
-        if (openapiSchema.definitions) {
-          for (const key in openapiSchema.definitions) {
-            // Preserve user-provided definitions
-            if (key in outputSchemas) continue;
-            const relatedSchema = openapiSchema.definitions[key];
-            debug(
-              '    defining referenced schema for %j: %j',
-              key,
-              relatedSchema,
-            );
-            outputSchemas[key] = relatedSchema;
-          }
-          delete openapiSchema.definitions;
-        }
-
-        debug('    defining schema for %j: %j', p.name, openapiSchema);
-        outputSchemas[p.name] = openapiSchema;
-        break;
+        generateOpenAPISchema(spec, p);
       }
     }
   }
   return spec;
 }
 
+/**
+ * Resolve the x-ts-type in schema objects
+ * @param spec Controller spec
+ * @param content Parent object of the schema
+ */
+function resolveTSType(
+  spec: ControllerSpec,
+  content: {[TS_TYPE_KEY]?: Function; schema?: SchemaObject},
+) {
+  debug('  evaluating content: %j', content);
+  const tsType = content[TS_TYPE_KEY];
+  debug('  %s => %o', TS_TYPE_KEY, tsType);
+  if (tsType) {
+    content.schema = resolveSchema(tsType, content.schema);
+    if (content.schema.$ref) generateOpenAPISchema(spec, tsType);
+
+    // We don't want a Function type in the final spec.
+    delete content[TS_TYPE_KEY];
+  } else {
```

Comment:
Let's return early so that the next block of code can be indented by one less level.